# BASELINE MODEL
-----------------
-----------------
-----------------

In [3]:
import pandas as pd
import numpy as np
import time
import gc
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Conv2D, Conv1D, GlobalMaxPooling2D, LSTM, MaxPooling2D, MaxPooling1D, Flatten
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.svm import SVC

In [4]:
!free -h

               total        used        free      shared  buff/cache   available
Mem:            15Gi       1.9Gi        12Gi       202Mi       911Mi        13Gi
Swap:          976Mi       671Mi       305Mi


In [3]:
# This df will be a row-concatenated version of the same channels
# baseline = pd.read_csv('../../baseline.csv', delimiter=',')
# baseline.drop('Unnamed: 0', inplace=True, axis=1)

In [4]:
# del df
gc.collect()

82

In [3]:
df = pd.read_csv('../data/dfTensor.csv', delimiter=';', encoding='latin-1')
df.drop('Unnamed: 0', inplace=True, axis=1)

In [7]:
df['F8'].iloc[0]

'[3989.23077, 3983.589744, 3987.692308, 3991.794872, 3992.307693, 3990.256411, 3993.333334, 4000.0, 3994.871795, 3985.128206, 3995.897436, 4001.538461, 3986.153847, 3982.051283, 3990.256411, 3994.358975, 3981.538462, 3961.025642, 3968.205129, 3989.23077, 3980.512821, 3973.333334, 3997.948718, 4009.230769, 3998.974359, 3998.461539, 3999.48718, 3997.435898, 3999.48718, 4001.025641, 3994.871795, 3987.179488, 3987.692308, 3992.307693, 3987.692308, 3980.0, 3984.102565, 3994.358975, 3991.794872, 3982.564103, 3980.0, 3978.461539, 3978.461539, 3989.23077, 3998.461539, 3998.461539, 3996.410257, 3978.974359, 3966.153847, 3978.974359, 3987.692308, 3985.128206, 3993.846154, 3995.384616, 3974.871795, 3957.948718, 3965.128206, 3990.769231, 4007.179487, 3998.974359, 3981.025642, 3970.256411, 3976.923077, 3993.333334, 3992.820513, 3978.461539, 3976.923077, 3975.897436, 3973.846154, 3984.615385, 3986.153847, 3978.974359, 3980.0, 3982.051283, 3981.538462, 3983.589744, 3978.974359, 3971.282052, 3972.8205

In [13]:
# df['O1'] = df['O1'].apply(eval)
# df['O1']

In [3]:
# loading dataframes for each channel, pick and choose at will

# occ0 = pd.read_csv('../data/occ0Exp.csv', delimiter=',')
# occ1 = pd.read_csv('../data/occ1Exp.csv', delimiter=',')
# fefF3 = pd.read_csv('../data/fefF3Exp.csv', delimiter=',')
# fefF4 = pd.read_csv('../data/fefF4Exp.csv', delimiter=',')
# fefF7 = pd.read_csv('../data/fefF7Exp.csv', delimiter=',')
# fefF8 = pd.read_csv('../data/fefF8Exp.csv', delimiter=',')
# temT7 = pd.read_csv('../data/temT7Exp.csv', delimiter=',')
# temT8 = pd.read_csv('../data/temT8Exp.csv', delimiter=',')
# pfcAF3 = pd.read_csv('../data/pfcAF3Exp.csv', delimiter=',')
# pfcAF4 = pd.read_csv('../data/pfcAF4Exp.csv', delimiter=',')
# motFC5 = pd.read_csv('../data/motFC5Exp.csv', delimiter=',')
# motFC6 = pd.read_csv('../data/motFC6Exp.csv', delimiter=',')
# parP7 = pd.read_csv('../data/parP7Exp.csv', delimiter=',')
# parP8 = pd.read_csv('../data/parP8Exp.csv', delimiter=',')

# loading same datasets expanded with filter processing

# occ0 = pd.read_csv('../data/occ0Proc.csv', delimiter=',')
# occ1 = pd.read_csv('../data/occ1Proc.csv', delimiter=',') 
# fefF3 = pd.read_csv('../data/fefF3Proc.csv', delimiter=',')
# fefF4 = pd.read_csv('../data/fefF4Proc.csv', delimiter=',')
# fefF7 = pd.read_csv('../data/fefF7Proc.csv', delimiter=',')
# fefF8 = pd.read_csv('../data/fefF8Proc.csv', delimiter=',') 
# temT7 = pd.read_csv('../data/temT7Proc.csv', delimiter=',')
# temT8 = pd.read_csv('../data/temT8Proc.csv', delimiter=',')
# pfcAF3 = pd.read_csv('../data/pfcAF3Proc.csv', delimiter=',')
# pfcAF4 = pd.read_csv('../data/pfcAF4Proc.csv', delimiter=',')
# motFC5 = pd.read_csv('../data/motFC5Proc.csv', delimiter=',')
# motFC6 = pd.read_csv('../data/motFC6Proc.csv', delimiter=',')
# parP7 = pd.read_csv('../data/parP7Proc.csv', delimiter=',')
# parP8 = pd.read_csv('../data/parP8Proc.csv', delimiter=',')

# occ0 = pd.read_csv('../data/occ0Wide.csv', delimiter=',')
# occ1 = pd.read_csv('../data/occ1Wide.csv', delimiter=',') 
# fefF3 = pd.read_csv('../data/fefF3Wide.csv', delimiter=',')
# fefF4 = pd.read_csv('../data/fefF4Wide.csv', delimiter=',')
# fefF7 = pd.read_csv('../data/fefF7Wide.csv', delimiter=',')
# fefF8 = pd.read_csv('../data/fefF8Wide.csv', delimiter=',') 
# temT7 = pd.read_csv('../data/temT7Wide.csv', delimiter=',')
# temT8 = pd.read_csv('../data/temT8Wide.csv', delimiter=',')
# pfcAF3 = pd.read_csv('../data/pfcAF3Wide.csv', delimiter=',')
# pfcAF4 = pd.read_csv('../data/pfcAF4Wide.csv', delimiter=',')
# motFC5 = pd.read_csv('../data/motFC5Wide.csv', delimiter=',')
# motFC6 = pd.read_csv('../data/motFC6Wide.csv', delimiter=',')
# parP7 = pd.read_csv('../data/parP7Wide.csv', delimiter=',')
# parP8 = pd.read_csv('../data/parP8Wide.csv', delimiter=',')

In [ ]:
# del occ0
# del occ1
# del fefF3
# del fefF4
# del fefF7
# del fefF8
# del temT7
# del temT8
# del pfcAF3
# del pfcAF4
# del motFC5
# del motFC6
# del parP7
# del parP8

# gc.collect()

In [5]:
# loading only the dataframes from channels positively correlated with code, to save RAM
# corrcols = [occ1, fefF4, fefF8, pfcAF4, motFC6, parP8] #removed motFC5 for being a diff shape

Correlations of averaged channels, if there is a need to pick and choose to save RAM:

<sub>
code          1.000000
    
&nbsp;
    
occ0Data     -0.000385

&nbsp;
    
occ1Data      0.001626
    
&nbsp;

fefF3Data    -0.005483

&nbsp;
    
fefF4Data     0.008288

&nbsp;
    
fefF7Data    -0.004799

&nbsp;
        
fefF8Data     0.006907

&nbsp;    
    
temT7Data    -0.002657

&nbsp;    
    
temT8Data    -0.001037

&nbsp;    
    
pfcAF3Data   -0.007512

&nbsp;    
    
pfcAF4Data    0.007591

&nbsp;    
    
motFC5Data    0.003757

&nbsp;    
    
motFC6Data    0.001372

&nbsp;    
    
parP7Data    -0.003000

&nbsp;    
    
parP8Data     0.002366
</sub>

In [8]:
!free -h

               total        used        free      shared  buff/cache   available
Mem:            15Gi       6.1Gi       1.8Gi       438Mi       7.6Gi       8.6Gi
Swap:          976Mi       352Mi       624Mi


In [7]:
# If individual columns are loaded, use this cell to process them and aggregate group them by event
# This gets rid of the time dimension though and will make the model be highly inaccurate

# dfs = [occ0, occ1, fefF3, fefF4, fefF7, fefF8, 
#        temT7, temT8, pfcAF3, pfcAF4, motFC5, motFC6, parP7, parP8]
# for i in dfs: #dfs:
# #     i.drop('Unnamed: 0.1', inplace=True, axis=1)
#     i.drop(['Unnamed: 0', 'size'], inplace=True, axis=1)
#     i['data'].astype(float, copy=False)
#     print(i.columns)
# #     i = i.groupby('event').mean()
# #     print(i.columns)

In [10]:
# non loop version of above's last line
# occ1, fefF4, fefF8, pfcAF4, motFC5, motFC6, parP8 for corrcols

# occ0 = occ0.groupby('event').mean()
# occ1 = occ1.groupby('event').mean()
# fefF3 = fefF3.groupby('event').mean()
# fefF4 = fefF4.groupby('event').mean()
# fefF7 = fefF7.groupby('event').mean()
# fefF8 = fefF8.groupby('event').mean()
# temT7 = temT7.groupby('event').mean()
# temT8 = temT8.groupby('event').mean()
# pfcAF3 = pfcAF3.groupby('event').mean()
# pfcAF4 = pfcAF4.groupby('event').mean()
# motFC5 = motFC5.groupby('event').mean() #this one has less rows than the rest, why?
# motFC6 = motFC6.groupby('event').mean()
# parP7 = parP7.groupby('event').mean()
# parP8 = parP8.groupby('event').mean()


In [8]:
## Check each channel dataframe for null vals 
# for i in corrcols:
#     print(max(i.isna().sum()))

In [58]:
## Renaming data column to specify channel name in individual channel dataframes, so when they are concatenated, the full df doesn't have multiple cols with the same names
## (use  occ1, fefF4, fefF8, pfcAF4, motFC6, parP8 for corrcols)

# occ0.rename(columns={'data': 'occ0data'}, inplace=True)
# occ1.rename(columns={'data':'occ1data', 'ICAdata':'occ1ICAdata', 'SGFdata':'occ1SGFdata'}, inplace=True)
# fefF3.rename(columns={'data': 'fefF3data'}, inplace=True)
# fefF4.rename(columns={'data':'fefF4data', 'ICAdata':'fefF4ICAdata', 'SGFdata':'fefF4SGFdata'}, inplace=True)
# fefF7.rename(columns={'data': 'fefF7data'}, inplace=True)
# fefF8.rename(columns={'data':'fefF8data', 'ICAdata':'fefF8ICAdata', 'SGFdata':'fefF8SGFdata'}, inplace=True)
# temT7.rename(columns={'data': 'temT7data'}, inplace=True)
# temT8.rename(columns={'data': 'temT8data'}, inplace=True)
# pfcAF3.rename(columns={'data': 'pfcAF3data'}, inplace=True)
# pfcAF4.rename(columns={'data':'pfcAF4data', 'ICAdata':'pfcAF4ICAdata', 'SGFdata':'pfcAF4SGFdata'}, inplace=True)
# motFC5.rename(columns={'data': 'motFC5data'}, inplace=True)
# motFC6.rename(columns={'data':'motFC6data', 'ICAdata':'motFC6ICAdata', 'SGFdata':'motFC6SGFdata'}, inplace=True)
# parP7.rename(columns={'data': 'parP7data'}, inplace=True)
# parP8.rename(columns={'data':'parP8data', 'ICAdata':'parP8ICAdata', 'SGFdata':'parP8SGFdata'}, inplace=True)


In [14]:
# occ0.columns, occ1.columns, fefF7.columns, fefF8.columns
# # dfs[0] #this wasn't changed with the above functions, it's only used for calling the original datasets as they were read in
# dfsProcessed = [occ0, occ1, fefF3, fefF4, fefF7, fefF8, 
#        temT7, temT8, pfcAF3, pfcAF4, motFC5, motFC6, parP7, parP8]

## Remaking this list included processed dataframes
# corrcolsProcessed = [occ1, fefF4, fefF8, pfcAF4, motFC6, parP8] #new vars because the original list has the unprocessed dataframes in memory it seems

In [15]:
## If using individual channel dataframes, use this cell to turn them into one df
# full = pd.concat(corrcolsProcessed, axis=1, join='inner')

In [18]:
# full.columns.duplicated()
# # Since the concat created duplicate code columns, using this trick to delete them:
# #https://stackoverflow.com/questions/14984119/python-pandas-remove-duplicate-columns
# #Although a different type of join might prevent the need for this
# full = full.loc[:,~full.columns.duplicated()].copy()
# #full = full.drop(['size', 'id' ], axis=1) # If any extraneous cols are leftover
# full.dtypes

In [9]:
# full

In [10]:
#corrMatFull = full.corr()
#corrMatFull['code']
# check = pd.read_csv('../data/occ1Exp.csv', delimiter=',')
# print(check.corr()) # code correlates to data in this channel by 0.001562 
# del check 

# check = pd.read_csv('../data/fefF7Exp.csv', delimiter=',')
# print(check.corr()) # code correlates to data in this channel by -0.004390.
# del check 

# check = pd.read_csv('../data/occ0Exp.csv', delimiter=',')
# print(check.corr()) # code correlates to data in this channel by -0.000308 
# del check 

# check = pd.read_csv('../data/fefF8Exp.csv', delimiter=',')
# print(check.corr()) # code correlates to data in this channel by 0.006184
# del check 

In [20]:
# full.plot(y=['occ0Data', 'occ1Data', 'fefF7Data', 'fefF8Data'], alpha=.7 ); 

In [25]:
!free -h

               total        used        free      shared  buff/cache   available
Mem:            15Gi       8.8Gi       4.4Gi       182Mi       2.3Gi       6.3Gi
Swap:          976Mi       381Mi       595Mi


--------------
## Baseline Model: Only O1, O2, F7, F8; No Preprocessing
## Baseline+ : Basline+Preprocessing

--------------

&nbsp;


In [183]:
## Setting up X and y, train-test split for "full"
# X = full.drop('code', axis=1)
# XExperiment = full.drop(['code', 'occ1data', 'fefF4data',
#                          'fefF8data', 'pfcAF4data', 'motFC6data', 'parP8data',
#                          'occ1SGFdata', 'fefF4SGFdata','fefF8SGFdata', 'pfcAF4SGFdata', #TODO change to pfc
#                          'motFC6SGFdata', 'parP8SGFdata'], axis=1) 
# y = full['code']
# print(X.columns, X.iloc[0], y.iloc[0], X.shape, y.shape)
# XTrain, XTest, yTrain, yTest = train_test_split(XExperiment, y)

In [4]:
# Setting up X and y
X = df.drop('code', axis=1)
y = df['code']

In [11]:
# del df
# gc.collect()

In [5]:
#Since the csv is formatted wrong, something with read_csv is encoding the lists as strings, run this to turn them back to lists
X = X.applymap(eval) # takes a while
print(X.iloc[0], type(y.iloc[0]), X.shape, y.shape)

O1     [4203.076923, 4193.333333, 4194.871794, 4207.1...
O2     [4229.743589, 4216.923076, 4207.179487, 4215.3...
F3     [4538.461538, 4528.717948, 4524.615384, 4526.1...
F4     [4682.051282, 4667.179487, 4662.051282, 4669.2...
F7     [4489.230769, 4475.384615, 4474.358974, 4486.6...
F8     [3989.23077, 3983.589744, 3987.692308, 3991.79...
T7     [4497.948717, 4498.461538, 4494.871794, 4497.9...
T8     [4506.666666, 4501.025641, 4496.923076, 4496.9...
AF3    [4395.384615, 4382.564102, 4377.435897, 4387.1...
AF4    [4078.461538, 4062.564102, 4055.897435, 4065.6...
FC5    [4207.692307, 4205.641025, 4200.51282, 4194.35...
FC6    [4227.692307, 4215.384615, 4210.76923, 4221.02...
P7     [4203.076923, 4192.820512, 4194.871794, 4204.6...
P8     [4245.641025, 4236.410256, 4218.461538, 4220.0...
Name: 0, dtype: object <class 'numpy.int64'> (65034, 14) (65034,)


In [6]:
# XO1 = pd.DataFrame(X['O1'].tolist())
# XO2 = pd.DataFrame(X['O2'].tolist())
# XF3 = pd.DataFrame(X['F3'].tolist())
# XF4 = pd.DataFrame(X['F4'].tolist())
# XF7 = pd.DataFrame(X['F7'].tolist())
# XF8 = pd.DataFrame(X['F8'].tolist())
# XT7 = pd.DataFrame(X['T7'].tolist())
# XT8 = pd.DataFrame(X['T8'].tolist())
# XAF3 = pd.DataFrame(X['AF3'].tolist())
# XAF4 = pd.DataFrame(X['AF4'].tolist())
# XFC5 = pd.DataFrame(X['FC5'].tolist())
# XFC6 = pd.DataFrame(X['FC6'].tolist())
# XP7 = pd.DataFrame(X['P7'].tolist())
# XP8 = pd.DataFrame(X['P8'].tolist())

In [20]:
#The columns are only labeled numerically, so there are 14 copies of the same set of names.
# I don't think this would be a problem, but if it is, a unique list of col names can
# be given when making each dataframe, at the X<channel name> = pd.DataFrame ... tolist(), cols =  
# XWide = pd.concat([XO1, XO2, XF3, XF4, XF7, XF8, XT7, XT8, XAF3,
#                    XAF4, XFC5, XFC6, XP7, XP8], axis=1)

In [79]:
# del X
# del XO1
# del XO2
# del XF3
# del XF4
# del XF7
# del XF8
# del XT7
# del XT8
# del XAF3
# del XAF4
# del XFC5
# del XFC6
# del XP7
# del XP8
gc.collect()

40

In [23]:
# XWide.to_csv('../data/XWide.csv', sep=',')
# y.to_csv('../data/y.csv', sep=',')

In [10]:
# Read pre-made X set, wide version
X = pd.read_csv('../data/XWide.csv').drop('Unnamed: 0', axis=1) 

In [11]:
y = pd.read_csv('../data/y.csv').drop('Unnamed: 0', axis=1) 

In [13]:
# train-test split for "df"
# XTrain, XTest, yTrain, yTest = train_test_split(X, y)
XTrain, XTest, yTrain, yTest = train_test_split(X, y)

In [14]:
del X
gc.collect()

118

In [15]:
XTrain

0            1            2            3            4  \
8533   4435.897435  4436.923076  4440.000000  4439.487179  4438.974358   
10641  4424.615384  4423.589743  4413.333333  4413.333333  4419.487179   
1189   4440.000000  4441.025641  4443.589743  4437.948717  4437.948717   
42297  4443.076923  4444.615384  4455.897435  4461.025641  4453.333333   
37216  4410.256410  4410.769230  4416.410256  4425.128205  4426.153846   
...            ...          ...          ...          ...          ...   
10808  4610.769230  4620.000000  4624.102564  4626.153846  4628.205128   
1804   4407.179487  4409.743589  4410.769230  4411.794871  4414.358974   
44105  4434.358974  4428.205128  4431.282051  4434.871794  4433.333333   
36503  4406.666666  4403.589743  4411.282051  4416.410256  4414.358974   
59551  4435.897435  4442.051282  4442.564102  4440.000000  4441.538461   

                 5            6            7            8            9  ...  \
8533   4441.025641  4440.512820  4438.974358  4443.076923  4447.692307  ...   
10641  4422.564102  4421.538461  4415.897435  4410.769230  4409.230769  ...   
1189   4443.076923  4436.410256  4425.128205  4429.230769  4437.435897  ...   
42297  4446.666666  4446.153846  4444.615384  4444.615384  4444.615384  ...   
37216  4414.358974  4405.641025  4407.179487  4412.307692  4420.512820  ...   
...            ...          ...          ...          ...          ...  ...   
10808  4633.333333  4640.512820  4645.641025  4652.307692  4662.564102  ...   
1804   4414.871794  4413.846153  4415.384615  4414.358974  4411.282051  ...   
44105  4430.256410  4427.692307  4428.205128  4428.205128  4427.692307  ...   
36503  4414.358974  4418.974358  4419.487179  4417.435897  4421.025641  ...   
59551  4442.051282  4441.538461  4447.692307  4454.358974  4457.435897  ...   

            246.13       247.13       248.13       249.13       250.13  \
8533   4195.384615  4198.461538  4194.871794  4194.358974  4202.564102   
10641  4152.307692  4153.846153  4156.923076  4161.538461  4162.051282   
1189   4257.948717  4250.769230  4240.512820  4238.974358  4241.538461   
42297  4235.384615  4227.692307  4218.461538  4218.974358  4225.128205   
37216  4224.615384  4222.051282  4210.769230  4202.564102  4207.179487   
...            ...          ...          ...          ...          ...   
10808  4020.512820  4028.717948  4031.282051  4015.384615  3998.461539   
1804   4242.564102  4234.358974  4230.769230  4231.794871  4227.179487   
44105  4230.769230  4233.333333  4237.435897  4245.128205  4244.102564   
36503  4204.615384  4207.179487  4215.897435  4218.461538  4213.846153   
59551  4069.743589  4065.128205  4078.461538  4090.256410  4089.230769   

            251.13       252.13       253.13       254.13       255.13  
8533   4214.358974  4215.384615  4207.179487  4203.076923  4202.564102  
10641  4158.461538  4163.589743  4169.743589  4164.615384  4162.564102  
1189   4246.666666  4251.794871  4248.717948  4245.128205  4249.230769  
42297  4225.128205  4223.076923  4225.128205  4227.692307  4227.692307  
37216  4215.384615  4218.461538  4210.256410  4198.974358  4204.102564  
...            ...          ...          ...          ...          ...  
10808  4007.692307  4028.717948  4034.358974  4028.205128  4023.589743  
1804   4229.230769  4246.153846  4244.102564  4223.076923  4218.461538  
44105  4235.897435  4233.846153  4237.435897  4232.820512  4221.538461  
36503  4210.769230  4212.820512  4216.923076  4216.923076  4211.794871  
59551  4087.692307  4084.615384  4076.410256  4073.333333  4073.333333  

[48775 rows x 3584 columns]

In [14]:
!cowsay moooooo... braaains...

 ________________________
< moooooo... braaains... >
 ------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


In [43]:
XTrain.shape, XTest.shape, yTrain.shape, yTest.shape

((48775, 256), (16259, 256), (48775,), (16259,))

In [14]:
# XTrainNP = XTrain.applymap(np.array)
# # type(XTrainNP['T7'][0])
# XTrainNP = XTrainNP.to_numpy()
# XTrainNP.shape

In [15]:
!free -h

               total        used        free      shared  buff/cache   available
Mem:            15Gi       6.9Gi       2.9Gi       238Mi       5.7Gi       8.1Gi
Swap:          976Mi       374Mi       602Mi


In [13]:
# XTrainNP = tf.reshape(np.array(XTrainNP), (48775, 256, 1, 14))
# XTrainNP = np.reshape(XTrainNP, newshape=(256, 1, 14))
ohe = OneHotEncoder(sparse=False)
# # to fix error, try this: https://www.youtube.com/watch?v=k_VAKyzggJI
# from sklearn.preprocessing import LabelEncoder
# lb = LabelEncoder()
# yTrainOHE = lb.fit_transform(yTrain)
yTrainOHE = ohe.fit_transform(yTrain.to_numpy().reshape(-1,1))

# ohe0 = OneHotEncoder(sparse=False)
# yTestOHE = ohe0.fit(yTrain.to_numpy().reshape(-1,1)) # Daniel said to do this but I can't remember why

In [45]:
yTrainOHE.shape

(48775, 11)

In [77]:
del model
gc.collect()

2733

In [52]:
model = Sequential()

In [ ]:
model.add(Conv2D(filters=64, kernel_size=1, activation='relu', input_shape=(48775, 256, 1)))
model.add(MaxPooling2D(pool_size=(1,1)))
model.add(Conv2D(filters=64, kernel_size=1, activation='relu'))
model.add(MaxPooling2D(pool_size=(1,1)))
model.add(Conv2D(filters=64, kernel_size=1, activation='relu'))
model.add(MaxPooling2D(pool_size=(1,1)))
model.add(Flatten())
# model.add(Dense(32, activation='relu'))
# model.add(Dense(32, activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(11, activation='softmax'))

# model.add(LSTM(7, input_shape=(48775, 256, 1), activation='tanh', 
#                recurrent_activation='sigmoid'))
# model.add(LSTM(7, input_shape=(48775, 256, 1), activation='tanh', 
#                recurrent_activation='sigmoid')
# model.add(LSTM(7, input_shape=(48775, 256, 1), activation='tanh', 
#                recurrent_activation='sigmoid')
## model.add(Dense(32, activation='relu'))
## model.add(Dense(32, activation='relu'))
# model.add(Dense(11, activation='softmax'))

In [49]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv1d_7 (Conv1D)           (None, 256, 64)           128       
                                                                 
 max_pooling1d_4 (MaxPooling  (None, 256, 64)          0         
 1D)                                                             
                                                                 
 conv1d_8 (Conv1D)           (None, 256, 64)           4160      
                                                                 
 max_pooling1d_5 (MaxPooling  (None, 256, 64)          0         
 1D)                                                             
                                                                 
 conv1d_9 (Conv1D)           (None, 256, 64)           4160      
                                                                 
 max_pooling1d_6 (MaxPooling  (None, 256, 64)         

In [ ]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])
results = model.fit(XTrain, yTrainOHE, batch_size=20,epochs=30, verbose=1)

Best accuracy for the unprocessed signal with all channels was about .1020

In 6/14 channels, adding SGF data was about the same

In 6/14 channels, adding SGF and dropping raw data resulted in really low accuracy and NaN loss

In 6/14 channels, ICA only had a accuracy of about .1000

6/14 channels, unproccessed+SGF+ICA, 22,059 params: NaN loss again, what's wrong with the data? woe be the futility of mine ways

6/14 channels, SGF only, 9,771 params, accuracy still around .1000

6/14 channels, raw, 9,771 params, accuracy about .1000, so the .0020 increase was only from the extra channels. Processing doesn't seem to change the accuracy at all.

----------
Using vector data rows ("wide"), I get this error:
```
UnimplementedError                        Traceback (most recent call last)
<ipython-input-59-f6344dc87bbc> in <module>
      1 model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])
----> 2 results = model.fit(XTrain, yTrainOHE, batch_size=20,epochs=30, verbose=1)

...

Cast string to float is not supported
	 [[{{node sequential_4/Cast}}]] [Op:__inference_train_function_2630]
```

but the prep function seems to work fine: column is series, col[0] is list, and col[0][0] is float
```
(learn-env) :~/flatiron/cap/capstoneProject/workingNotebooks$ python dataPrepV2.py 
<class 'pandas.core.series.Series'> <class 'list'> <class 'float'>
```

-------------

Using one channel with each data row expanded as a list then made into a new df, the loss gave a NaN error and the accuracy hovered around 0.002 (Conv1D)

Attempting a Conv2D version with input shape 48775, 256, 1 gives the following memory overload error:

```
ResourceExhaustedError: {{function_node __wrapped__StatelessRandomUniformV2_device_/job:localhost/replica:0/task:0/device:CPU:0}} OOM when allocating tensor with shape[799129600,32] and type float on /job:localhost/replica:0/task:0/device:CPU:0 by allocator cpu [Op:StatelessRandomUniformV2]
```

I believe the temporal + amplitude data in the raw readings would be maintained using Conv2D layers, but these data trigger an error in TF.

### Support Vector Machine Attempt

In [18]:
svc = SVC()
svc.fit(XTrain.fillna(XTrain.mean()), yTrain)

/home/h/anaconda3/envs/learn-env/lib/python3.8/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


SVC()

In [19]:
svc.score(XTest.fillna(XTest.mean()), yTest)

0.11212251675994833

A quick support vector classifier ended up getting the same accuracy, 0.10234331754720463

Another with the one channel "wide" format gives an accuracy of 0.10910880127929147, which is interesting because it's better than the baseline CNN, and it's 1 out of 14 channels unfiltered. Maybe a not-so-bad model will be a possibility with all channels, especially if they're filtered.

On the full, 14 channel dataset, I got my highest score yet, 0.11212251675994833! now we're getting somewhere lol. Tomorrow I will try the filtered version 

### Echo State Network Attempt

In [ ]:
import reservoirpy as rpy
from reservoirpy.nodes import Reservoir, Ridge, FORCE, ESN

readout = Ridge(ridge=1e-7)
train_states = Reservoir.run(XTrain, reset=True)
# train_statesy = reservoir.run(yTrain.to_numpy().reshape(-1, 1), reset=True)
readout = readout.fit(train_states, yTrain.to_numpy().reshape(-1, 1), warmup=10)
test_states = reservoir.run(XTest.to_numpy().reshape(-1, 1))
yPred = readout.run(test_states)

reservoir = Reservoir(100, lr=0.5, sr=0.9)
ridge = Ridge(ridge=1e-7)

esn_model = reservoir >> ridge

esn_model = esn_model.fit(XTrain.to_numpy().reshape(-1, 1), yTrain.to_numpy().reshape(-1, 1).to_numpy().reshape(-1, 1), warmup=10)
print(reservoir.is_initialized, readout.is_initialized, readout.fitted)

yPred = esn_model.run(XTest.to_numpy().reshape(-1, 1))


While it was nice to learn how to set one of these up, I realized after a lot of tuning the reservoirpy library doesn't seem to have good interop with pandas. Might have interesting results with 1D timestamp and 1D value arrays, but not a multidimensional dataframe

In [ ]:
readout1 = Ridge(ridge=1e-7)
train_states1 = reservoir.run(full['pfcAF4SGFdata'].to_numpy().reshape(-1, 1), reset=True)
# train_statesy = reservoir.run(yTrain.to_numpy().reshape(-1, 1), reset=True)
readout1 = readout1.fit(train_states1, full['code'].to_numpy().reshape(-1, 1), warmup=10)
test_states1 = reservoir.run((full['pfcAF4SGFdata'].to_numpy().reshape(-1, 1)))
# yPred1 = readout.run(test_states1)

reservoir1 = Reservoir(100, lr=0.5, sr=0.9)
ridge1 = Ridge(ridge=1e-7)

esn_model1 = reservoir >> ridge

esn_model1 = esn_model1.fit(full['pfcAF4SGFdata'].to_numpy().reshape(-1, 1), 
                             full['code'].to_numpy().reshape(-1, 1) ,warmup=10)
print(reservoir.is_initialized, readout.is_initialized, readout.fitted)

yPred1 = esn_model.run((full['pfcAF4SGFdata'].to_numpy().reshape(-1, 1)))

In [220]:
set(yPred1.flatten()), (mean())

{4.494217014863551}

In [ ]:
# TODO use exploded data of one event from original CSV, predict brain activity with ESN:
# forecast, generate artificial brainwaves of looking at a digit

# TODO explore possibility of using matplotlib plots of signals as training data
# like in https://arxiv.org/abs/1511.06448

# in the future, try iterating ESNpredict -> ESNtrain -> ESNpredict -> ESNtrain -> ...
# to see how predictions change over time, maybe there would be a "core" shape that can be
# like a waveform decomposition of the number, 
